Fix/issue 12944 member cant leave the community

MishkaRogachev opened the pull request:
### What does the PR do
Fix an issue i've introduced here: https://github.com/status-im/status-desktop/pull/12891
Also fixes https://github.com/status-im/status-desktop/issues/12944

Before sending the notifications to the AC service, check if they are present in the response

### Affected areas

Communities, Contacts, Messages, Activity Center

### Video

https://github.com/status-im/status-desktop/assets/2522130/4ea7b307-a5e1-40b2-8b2a-1851ee1fccf0

caybro reviewed (commented):
Uhm, too much identical and repeated code :) Would benefit from some refactoring

MishkaRogachev commented:
> Uhm, too much identical and repeated code :) Would benefit from some refactoring

Refactored to the free function, to avoid response object copy

caybro reviewed (approved):
Looks cleaner but still.. isn't there a single, more central place to handle that?

jakubgs commented:
The build failures on `windows-02` were most probably due to corrupted Git repo. I cleared the workspaces and restarted it.

kounkou reviewed (changes requested):
Hi,
I first started checking the behavior before reviewing the code.
Currently, there's a crash anytime I test the following behavior. Please have a look at the video

Also in the log I can see something related to `onMarkAllMessagesRead` and `AsyncMarkAllMessagesReadTaskArg`: 

```
DBG 2023-12-07 23:44:03.003-08:00 [threadpool task thread] initiating task   topics="task-threadpool" tid=1292108 file=threadpool.nim:54 messageType=AsyncMarkAllMessagesReadTaskArg:ObjectType threadid=1292108 task="{\"$type\":\"AsyncMarkAllMessagesReadTaskArg:ObjectType\",\"chatId\":\"0x04c429330e366166687319c372775b39a917352e1af7f1510bf0a7a7805f7a2597030753529f3ff5e55b5a27c1d0be154312fa924532ac2ad005ed403223a43e9b\",\"vptr\":94387796220736,\"slot\":\"onMarkAllMessagesRead\",\"tptr\":94387764782064}"
DBG 2023-12-07 23:44:03.004-08:00 NewBE_callPrivateRPC                       topics="rpc" tid=1292108 file=core.nim:27 rpc_method=wakuext_markAllRead
Traceback (most recent call last)
/home/jacques/Documents/work/status-desktop/vendor/nim-taskpools/taskpools/taskpools.nim(171) workerEntryFn
/home/jacques/Documents/work/status-desktop/vendor/nim-taskpools/taskpools/taskpools.nim(242) eventLoop
/home/jacques/Documents/work/status-desktop/vendor/nim-taskpools/taskpools/taskpools.nim(193) runTask
/home/jacques/Documents/work/status-desktop/vendor/nimbus-build-system/vendor/Nim/lib/std/tasks.nim(79) invoke
/home/jacques/Documents/work/status-desktop/vendor/nimbus-build-system/vendor/Nim/lib/std/tasks.nim(121) runTask
/home/jacques/Documents/work/status-desktop/src/app/core/tasks/threadpool.nim(60) runTask
/home/jacques/Documents/work/status-desktop/src/app_service/service/message/async_tasks.nim(136) :anonymous
/home/jacques/Documents/work/status-desktop/vendor/nim-serialization/serialization.nim(20) finish
/home/jacques/Documents/work/status-desktop/vendor/nim-json-serialization/json_serialization/writer.nim(169) writeValue
/home/jacques/Documents/work/status-desktop/vendor/nimbus-build-system/vendor/Nim/lib/pure/json.nim(820) $
/home/jacques/Documents/work/status-desktop/vendor/nimbus-build-system/vendor/Nim/lib/pure/json.nim(718) toUgly
/home/jacques/Documents/work/status-desktop/vendor/nimbus-build-system/vendor/Nim/lib/pure/json.nim toUgly
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
make: *** [Makefile:815: run-linux] Segmentation fault (core dumped)

```

[crash-when-read-all.webm](https://github.com/status-im/status-desktop/assets/2589171/f1cc9dc3-3cb5-4e72-9470-f776174cb519)

MishkaRogachev commented:
> Hi, I first started checking the behavior before reviewing the code. Currently, there's a crash anytime I test the following behavior. Please have a look at the video
> 
Thanks for checking. I will do some more tests, because simple way to mark as read in 1-to-1 chat works for me:

https://github.com/status-im/status-desktop/assets/2522130/d13a75c5-8cda-434a-9445-673bcf8bb721

MishkaRogachev commented:
@kounkou fixed & rebased

https://github.com/status-im/status-desktop/assets/2522130/52dbaecd-5757-430f-b15f-ef5bde9eba2c

kounkou commented:
> @kounkou fixed & rebased
> 
>  Screen.Recording.2023-12-11.at.19.47.23.mov

Amazing, thanks, checked on my side, and the crash another crash is now happening when freshly opening a conversation. You might also need to fix conflicts and re-test ?
Please let me know if I can help in anyways.

status-im-auto commented:
#### :heavy_check_mark: [status-desktop/prs/linux/x86_64/tests-nim/PR-12958#13](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/tests-nim/job/PR-12958/13/) :small_blue_diamond: ~6 min 44 sec :small_blue_diamond: 3d77de03 :small_blue_diamond: [:package: tests/nim package](https://ci.status.im/job/status-desktop/job/prs/job/linux/job/x86_64/job/tests-nim/job/PR-12958/13//consoleText)

MishkaRogachev commented:
@kounkou i've rebased and cleaned the PR. Have run a few tests and don't see any more crashes. Btw, some crashes was because of this line: https://github.com/status-im/status-desktop/pull/12958/files#diff-ae9d357375682781e8d083507420d6c68aa6397b61f47be7f15141cd948a6ad5R105. nim can't iterate through empty json node

kounkou commented:
> ### Some checks haven’t completed yet
> 1 pending and 8 successful checks


Hi there,

I did : 

```
git checkout origin/fix/issue-12944-member-cant-leave-the-community
git reset --hard origin/fix/issue-12944-member-cant-leave-the-community
```

and tried to do a mark as unread to check for regression, and I got a crash on this branch.
To make sure, I went back on master, and it works perfect on master
So I couldn't test further than this.
Still need some work.

logs : 

```
DBG 2023-12-12 12:00:41.710-06:00 NewBE_callPrivateRPC                       topics="rpc" tid=86087 file=core.nim:27 rpc_method=wakuext_slowdownArchivesImport
DBG 2023-12-12 12:00:44.875-06:00 [threadpool task thread] initiating task   topics="task-threadpool" tid=86115 file=threadpool.nim:54 messageType=AsyncMarkMessageAsUnreadTaskArg:ObjectType threadid=86115 task="{\"$type\":\"AsyncMarkMessageAsUnreadTaskArg:ObjectType\",\"messageId\":\"0xe32e24eaa6838b72dc441d553ead68e7e7f9309c013e70204ec75a20e052ac9b\",\"chatId\":\"0x04c429330e366166687319c372775b39a917352e1af7f1510bf0a7a7805f7a2597030753529f3ff5e55b5a27c1d0be154312fa924532ac2ad005ed403223a43e9b\",\"vptr\":93829465742448,\"slot\":\"onAsyncMarkMessageAsUnread\",\"tptr\":93829411709728}"
DBG 2023-12-12 12:00:44.875-06:00 NewBE_callPrivateRPC                       topics="rpc" tid=86115 file=core.nim:27 rpc_method=wakuext_markMessageAsUnread
Traceback (most recent call last)
/home/jacques/Documents/w/status-desktop/vendor/nim-taskpools/taskpools/taskpools.nim(171) workerEntryFn
/home/jacques/Documents/w/status-desktop/vendor/nim-taskpools/taskpools/taskpools.nim(242) eventLoop
/home/jacques/Documents/w/status-desktop/vendor/nim-taskpools/taskpools/taskpools.nim(193) runTask
/home/jacques/Documents/w/status-desktop/vendor/nimbus-build-system/vendor/Nim/lib/std/tasks.nim(79) invoke
/home/jacques/Documents/w/status-desktop/vendor/nimbus-build-system/vendor/Nim/lib/std/tasks.nim(121) runTask
/home/jacques/Documents/w/status-desktop/src/app/core/tasks/threadpool.nim(60) runTask
/home/jacques/Documents/w/status-desktop/src/app_service/service/message/async_tasks.nim(334) :anonymous
/home/jacques/Documents/w/status-desktop/vendor/nim-serialization/serialization.nim(20) finish
/home/jacques/Documents/w/status-desktop/vendor/nim-json-serialization/json_serialization/writer.nim(169) writeValue
/home/jacques/Documents/w/status-desktop/vendor/nimbus-build-system/vendor/Nim/lib/pure/json.nim(820) $
/home/jacques/Documents/w/status-desktop/vendor/nimbus-build-system/vendor/Nim/lib/pure/json.nim(718) toUgly
/home/jacques/Documents/w/status-desktop/vendor/nimbus-build-system/vendor/Nim/lib/pure/json.nim toUgly
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
make: *** [Makefile:815: run-linux] Segmentation fault (core dumped)
```

kounkou commented on src/app_service/common/activity_center.nim:
```diff
@@ -1,8 +1,16 @@
   activityCenterNotifications*: JsonNode
+
+proc checkAndEmitACNotificationsFromResponse*(events: EventEmitter, result: JsonNode) =
+  if result{"activityCenterNotifications"} == nil or result{"activityCenterNotifications"}.kind == JNull:
```
Could be good to simplify this expression to avoid retyping result{"activityCenterNotification"}, maybe using the get(...) and avoid comparing to nil.

MishkaRogachev replied:
Good advice and it made me think that it's better to pass `result{"activityCenterNotifications"}` to the function, since we can't be sure yet that all responses will have notifications by that name. If you have concerns about this approach, let's open a ticket and discuss it, and I think there is still a lot we can fix. 

About nil and JNull checks. This procedure is used in several places, we have no control because AC notifications here can be either nil or null-object (even if this response is constructed from status-go json).

kounkou commented on src/app_service/service/message/async_tasks.nim:
```diff
@@ -125,19 +125,16 @@ const asyncMarkAllMessagesReadTask: Task = proc(argEncoded: string) {.gcsafe, ni
   let response =  status_go.markAllMessagesFromChatWithIdAsRead(arg.chatId)
 
-  var activityCenterNotifications: JsonNode
+  var activityCenterNotifications: JsonNode = newJObject()
```
Is it the case that we want to always pass the activityCenterNotifications object even when it's empty? why is it so? I thought that the field can be nil hence not added to the responseJson making the AC notification processing component less time. Can you tell me more why we are always forcing the AC notifications to be present please?

MishkaRogachev replied:
In this particular case, the choice between a null value or a null object (pattern) is a matter of taste (). I chose the null object for reasons:
- this is the approach elsewhere in the project (a very bad reason :) )
- it seems safer to me, since the code below crashes:
```
var nullJson: JsonNode # missing `= newJObject()` here
    if nullJson.kind == JNull:
      ...
```
- and there is no need to do a check.

kounkou reviewed (approved):
I wouldn't block on my comments, but I think it could be good to comment on why we are taking some decision in the code, rather than other decisions...
